### PR TITLE
[Chore]: Technical: Translate layer factory

### PR DESCRIPTION
**Pull request by HeimEndyd**

### CODE REVIEW

File: src/constants/color-ranges.ts [hunk 1]
```diff
@@ -22,6 +22,15 @@ import colorbrewer from 'colorbrewer';
 import {HexColor} from 'reducers';
 import {VizColorPalette} from './custom-color-ranges';
 
+export type ColorRange = {
+  name: string;
+  type: string;
+  category: string;
+  colors: HexColor[];
+  reversed?: boolean;
+  colorMap?: Map<string, HexColor> | object;
```
Comment (heshan0131) [inline src/constants/color-ranges.ts:31]: add 
```js
colorLegends?: Record<HexColor, string>;
```

File: src/layers/base-layer.ts [hunk 2]
```diff
@@ -182,8 +180,7 @@ class Layer<LayerConfig extends LayerBaseConfig = LayerBaseConfig> {
   id: string;
   // TODO: define meta
   meta: {};
-  // TODO: define visConfigSettings
-  visConfigSettings: {};
+  visConfigSettings: Partial<LayerVisConfigSettings>;
```
Comment (heshan0131) [inline src/layers/base-layer.ts:183]: Each layer has different propertues in visConfig and visConfigSettings. We should make the type here more generic and add specific types for individual layer's `layer.visConfigSettings`

```js
visConfigSettings: Record<string, VisConfig>;
```

Comment (heshan0131) [reply src/layers/base-layer.ts:183]: then in ArcLayer:

```
type ArcLayerVisConfigSettings = {
  opacity: VisConfigNumber;
  thickness: VisConfigNumber;
  colorRange: VisConfigColorRange;
  sizeRange: VisConfigRange;
  targetColor: VisConfigColorSelect;
};

export default class ArcLayer extends Layer {
  visConfigSettings: ArcLayerVisConfigSettings;
...
}
```

Comment (HeimEndyd) [reply src/layers/base-layer.ts]: I have considered your suggestions. This is how I see the implementation of ArcLayer

`type ArcLayerVisConfigSettings = {
  opacity: VisConfigNumber;
  thickness: VisConfigNumber;
  colorRange: VisConfigColorRange;
  sizeRange: VisConfigRange;
  targetColor: VisConfigColorSelect;
};

export default class ArcLayer extends Layer {
  declare visConfigSettings: ArcLayerVisConfigSettings;
...
}`

File: src/layers/base-layer.ts [hunk 3]
```diff
@@ -178,13 +176,14 @@ function* generateColor(): Generator<RGBColor> {
...
-class Layer<LayerConfig extends LayerBaseConfig = LayerBaseConfig> {
+class Layer {
   id: string;
   // TODO: define meta
   meta: {};
-  // TODO: define visConfigSettings
-  visConfigSettings: {};
-  config: LayerConfig;
+  visConfigSettings: {
+    [key in keyof Partial<LayerVisConfig>]: ValueOf<LayerVisConfigSettings>;
```
Comment (heshan0131) [inline src/layers/base-layer.ts:184]: Can we just put `[key: string]` here? people can create custom layers and assign custom visConfig props that is not part of the standard `LayerVisConfig`. This should be allowed by BaseLayer

**Pull request merged**